### DISCUSSION TITLE
on esp32 cam board using mongoose as client to send websocket data to server,too slow

### DISCUSSION THREAD

**Discussion by mnicnt** [cesanta/mongoose#1820] [Ask questions, get answers] [answered]
hi, I just on esp32 cam board using mongoose as websocket client to send mjpeg data to server, about 10k data a time,  but it sent too slow to a websocket server . about sending less 10k data,used >100 ms ,    I also tested that from a linux platform ,also using mongoose as websocket client , to send the same data , its normal fast  .  I am confusing that .  Is anyone meet it too ,how to solve it ? thanks!
on ESP32  , I use arduino IDE to develop , code like this:

..
static void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
  if (ev == MG_EV_OPEN) {
...
  } else if (ev == MG_EV_WS_OPEN) {
        c->label[0]='W';
  }
...
}

...
static void timer_fn(void *arg) {
  struct mg_mgr *mgr = (struct mg_mgr *) arg;
  for (struct mg_connection *c = mgr->conns; c != NULL; c = c->next) {
      if(c->label[0]=='W')
    {
      // have websocket connectd 
     mg_ws_send(c,  fb->buf, fb->len, WEBSOCKET_OP_BINARY) ; //   data about 10k bytes
    }
  }//for
} //timer_fn
...
void setup(){
 mg_mgr_init(&mgr);
 delay(3000);

 mg_timer_add(&mgr, 20 , MG_TIMER_REPEAT, timer_fn, &mgr);
 mg_ws_connect(&mgr, s_url, fn, &done,...
...
}

void loop() {
  mg_mgr_poll(&mgr, 10);
}

**Comment by scaprile** [comment, marked as answer]
Unless you absolutely need to broadcast the same data to several consumers (and even so, MQTT would probably be faster by just sending it once...) I would use the connection event handler and send data on `MG_EV_WRITE`
https://mongoose.ws/documentation/#events
Besides that, you may be having WiFi (RF) issues.

**Comment by mnicnt** [comment]
@scaprile thanks very much !
I only need to send one consumer(act as a server).  i have tested using mqtt ( beginpublish -> write ->endpublish) ,is faster than using mg_ws_send.   but how to code it by using  MG_EV_WRITE  event ? can you give me an example?  thanks !

**Reply by scaprile** [reply]
That event will fire once some data has been sent. If you can put all your data in one call, it will be sent and you don't need it. If you can't, ev_data will tell you how many bytes have been sent, move your pointer and send another chunk of data. This way you can stream a constant flow of data.

**Reply by mnicnt** [reply]
@scaprile thanks for your reply.
I put al data in one call.  the problem is  use MG_WS_SEND ,i found is too slow .  i tested useing mqtt ,it is ok ( in mqtt ,i used one by one small packages to send  by using beginpublish endpublish )  . i also tested from another linux based machine , do the same work like using esp32 cam board , it is also ok .  I want to know which functions can use to replace MG_WS_SEND ,  like use beginpublish/endpublish  , instead of  publishclient->publish .  thks!

**Reply by scaprile** [reply]
https://mongoose.ws/tutorials/

**Reply by mnicnt** [reply]
@scaprile  thks for ur reply!
